components/tor/tor_file_watcher.cc

```diff
@@ -0,0 +1,283 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/tor/tor_file_watcher.h"
+
+#include <string>
+#include <utility>
+
+#include "base/files/file.h"
+#include "base/sequenced_task_runner.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/task/post_task.h"
+#include "base/task/task_traits.h"
+#include "base/time/time.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+
+namespace tor {
+
+namespace {
+
+constexpr base::TaskTraits kWatchTaskTraits = {
+    base::ThreadPool(), base::MayBlock(), base::TaskPriority::BEST_EFFORT};
+
+#if defined(OS_WIN)
+constexpr char kControlPortMinTmpl[] = "PORT=1.1.1.1:1\r\n";
+constexpr char kControlPortMaxTmpl[] = "PORT=255.255.255.255:65535\r\n";
+constexpr char kLineBreak[] = "\r\n";
+#else
+constexpr char kControlPortMinTmpl[] = "PORT=1.1.1.1:1\n";
+constexpr char kControlPortMaxTmpl[] = "PORT=255.255.255.255:65535\n";
+constexpr char kLineBreak[] = "\n";
+#endif
+constexpr char kControlAuthCookieName[] = "control_auth_cookie";
+constexpr char kControlPortName[] = "controlport";
+}  // namespace
+
+TorFileWatcher::TorFileWatcher(const base::FilePath& watch_dir_path)
+    : polling_(false),
+      repoll_(false),
+      watch_dir_path_(std::move(watch_dir_path)),
+      watch_task_runner_(
+          base::ThreadPool::CreateSequencedTaskRunner(kWatchTaskTraits)),
+      watcher_(new base::FilePathWatcher,
+               base::OnTaskRunnerDeleter(watch_task_runner_)) {
+  DETACH_FROM_SEQUENCE(watch_sequence_checker_);
+}
+
+TorFileWatcher::~TorFileWatcher() {}
+
+void TorFileWatcher::StartWatching(WatchCallback callback) {
+  watch_callback_ = std::move(callback);
```

Comment:
basically there are only few methods that don't work on the task runner, so it is basically single-threaded class and we can hold it using `unique_ptr` and post using weak pointers. This will be better and simpler design (basically, extracting it from `TorControl` already is a very good step forward)

Reply:
cannot use weak pointer here because it will lead to invalidate the weak pointer on different sequence. All the watcher's tasks are done in a different thread `watch_task_runner_`

Reply:
fixed in https://github.com/brave/brave-core/pull/7927/commits/1d55c4b4a1b23d887344a68fad9dc153716f95be